File: LeetCode/1-two_sum.py
def two_sum_v1(nums, target):
    lut = {}
    for i in range(len(nums)):
        if nums[i] in lut:
            return [lut[nums[i]], i]
        else:
            lut[target - nums[i]] = i

def two_sum_v2(nums, target):
    n = len(nums)
    index = list(range(n))
    index.sort(key = lambda i: nums[i])
    l, r = 0, n-1
    while l < r:
        i, j = index[l], index[r]
        if nums[i] + nums[j] == target:
            return [i, j]
        elif nums[i] + nums[j] < target:
            l += 1
        else: # nums[i] + nums[j] > target
            r -= 1

class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        return two_sum_v2(nums, target)

File: LeetCode/1-two_sum.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    template<typename T, typename F>
    static int binary_seatch(const T* arr, int n, const T& key, F mapper) {
        int l = 0, r = n;
        while (l < r) {
            int m = (l + r) >> 1;
            const T& val = mapper(arr[m]);
            if (val == key) {
                return m;
            } else if (val < key) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return -1;
    }

    std::vector<int> twoSum(const std::vector<int>& nums, int target)
    {
        const int n = nums.size();

        std::vector<int> indirect(n);
        for (int i = 0; i < n; ++i)
            indirect[i] = i;

        std::sort(indirect.begin(), indirect.end(), [&](const int& a, const int& b)->bool{ return nums[a] < nums[b]; });

        for (int i = 0; i < n; ++i) {
            int j = binary_seatch(indirect.data(), n, target - nums[indirect[i]], [&](int k)->int{ return nums[k]; });
            if (j != -1 && j != i) {
                std::vector<int> result;
                result.reserve(2);
                result.push_back(indirect[i]);
                result.push_back(indirect[j]);
                return result;
            }
        }

        return std::vector<int>();
    }
};
